slick-hikaricp/src/main/scala/slick/jdbc/hikaricp/HikariCPJdbcDataSource.scala

```diff
@@ -1,16 +1,66 @@
 package slick.jdbc.hikaricp
 
-import java.sql.{Driver, Connection}
+import java.sql.{Connection, Driver}
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.ReentrantLock
+
 import com.typesafe.config.Config
 import slick.SlickException
-import slick.jdbc.{JdbcDataSourceFactory, JdbcDataSource}
+import slick.jdbc.{JdbcDataSource, JdbcDataSourceFactory}
 import slick.util.ConfigExtensionMethods._
+import slick.util.Logging
 
 /** A JdbcDataSource for a HikariCP connection pool.
   * See `slick.jdbc.JdbcBackend#Database.forConfig` for documentation on the config parameters. */
-class HikariCPJdbcDataSource(val ds: com.zaxxer.hikari.HikariDataSource, val hconf: com.zaxxer.hikari.HikariConfig) extends JdbcDataSource {
-  def createConnection(): Connection = ds.getConnection()
+class HikariCPJdbcDataSource(val ds: com.zaxxer.hikari.HikariDataSource, val hconf: com.zaxxer.hikari.HikariConfig)
+  extends JdbcDataSource
+  with Logging {
+
+  private[this] val poolLock: ReentrantLock = new ReentrantLock()
+  private[this] val condition = poolLock.newCondition()
+  private[this] var count: Int = 0
+
+  def createConnection(): Connection = {
+    poolLock.lock()
+    try {
+      if (logger.isDebugEnabled) logger.debug("number of connections in use = "+count)
+      while (count == hconf.getMaximumPoolSize) {
+          condition.await()
+      }
+      val c = new DelegateConnection(ds.getConnection()) {
+        override def close(): Unit = {
+          closeConnection(conn)
```

Comment:
super.close() closes the delegated connection, which will be closed in the closeConnection method. No need to close it twice.
I could refactor this and use super.close() but then the code in closeConnection would have to move also:

      val c = new DelegateConnection(ds.getConnection()) {
        override def close(): Unit = {
          poolLock.lock()
          try {
            super.close()
          } finally {
            connectionsInUseCount -= 1
            if (connectionsInUseCount == hconf.getMaximumPoolSize -1) {
              logger.debug("Connection freed below maximum, resuming")
              poolNotFullCondition.signal()
              resume()
            }
            poolLock.unlock()
          }
        }
      }

Is this preferred? Maybe made use an explicitly named inner class iso anonymous?